### PR TITLE
Pipeline: support fine grained shuffle

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #6518

Problem Summary:

### What is changed and how it works?
- refine `void buildPipelineExec(PipelineExecGroupBuilder & group_builder, Context & context, size_t concurrency);` to `void buildPipelineExec(PipelineExecutorStatus & exec_status, PipelineExecGroupBuilder & group_builder, Context & context, size_t concurrency);` and `PipelineExecGroupBuilder` no longer held `PipelineExecutorStatus`.
- Add function `getFineGrainedShuffle` of `PhysicalPlanNode` and then `Pipeline` will use this function to determine whether the execution mode is fine-grained or non-fine-grained.
- Add `FineGrainedPipelineEvent` to support fine grained execution mode and remove useless `PipelineEvent`.
```
 *        ┌──non fine grained mode──┐                          ┌──fine grained mode──┐
 *                                            ┌──FineGrainedPipelineEvent◄───FineGrainedPipelineEvent
 * PlainPipelineEvent◄───PlainPipelineEvent◄──┼──FineGrainedPipelineEvent◄───FineGrainedPipelineEvent
 *                                            ├──FineGrainedPipelineEvent◄───FineGrainedPipelineEvent
 *                                            └──FineGrainedPipelineEvent◄───FineGrainedPipelineEvent
```
- support fine grained shuffle in exchange receiver/sender.
- add fullstack test for pipeline mode

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Pipeline/Pipeline.cpp
Comment (ywqzzy) [inline dbms/src/Flash/Pipeline/Pipeline.cpp:77]: Then the next pipeline's FineGrainedPipelineEvent must relay on all FineGrainedPipelineEvents before this pipeline?

File: dbms/src/Flash/Planner/Plans/PhysicalExchangeSender.cpp
Comment (ywqzzy) [inline dbms/src/Flash/Planner/Plans/PhysicalExchangeSender.cpp:99]: Define the magic number 1024

File: dbms/src/Flash/Pipeline/Pipeline.cpp
Comment (SeaRise) [reply dbms/src/Flash/Pipeline/Pipeline.cpp:77]: yes

File: dbms/src/Flash/Planner/Plans/PhysicalExchangeSender.cpp
Comment (SeaRise) [reply dbms/src/Flash/Planner/Plans/PhysicalExchangeSender.cpp]: ok, updated,

File: dbms/src/Flash/Pipeline/Pipeline.cpp
Comment (ywqzzy) [reply dbms/src/Flash/Pipeline/Pipeline.cpp:77]: Maybe add some log here to identify different scene.

**Review by ywqzzy — APPROVED** [review]

File: dbms/src/Flash/Pipeline/Pipeline.cpp
Comment (SeaRise) [reply dbms/src/Flash/Pipeline/Pipeline.cpp:77]: Adding logs here is meaningless now, because there is only one fine grained pipeline at most...

File: dbms/src/Flash/Pipeline/Pipeline.cpp
Comment (yibin87) [inline dbms/src/Flash/Pipeline/Pipeline.cpp:41]: A little strange condition branch here, seems not that straightforward, can we take fine grained info directly into consideration here?

File: dbms/src/Flash/Pipeline/Schedule/Events/FineGrainedPipelineEvent.cpp
Comment (yibin87) [inline dbms/src/Flash/Pipeline/Schedule/Events/FineGrainedPipelineEvent.cpp:23]: For just one element vector, reserve seems not efficient.

File: dbms/src/Flash/Pipeline/Schedule/Events/PlainPipelineEvent.cpp
Comment (yibin87) [inline dbms/src/Flash/Pipeline/Schedule/Events/PlainPipelineEvent.cpp:24]: Not sure if we can use const auto & pipeline_exec_group to avoid some copy work

File: dbms/src/Flash/Pipeline/Schedule/Events/FineGrainedPipelineEvent.cpp
Comment (SeaRise) [reply dbms/src/Flash/Pipeline/Schedule/Events/FineGrainedPipelineEvent.cpp]: ok, removed.

File: dbms/src/Flash/Pipeline/Pipeline.cpp
Comment (SeaRise) [reply dbms/src/Flash/Pipeline/Pipeline.cpp]: ok, refined as
```
struct PipelineEvents
{
    Events events;
    bool is_fine_grained;

    PipelineEvents(Events && events_, bool is_fine_grained_);

    void mapInputs(const PipelineEvents & inputs);
};
```

**Review by yibin87 — APPROVED** [review]
LGTM， better merge it after code freeze:)

**Comment by SeaRise** [comment]
> LGTM， better merge it after code freeze:)

Okay

**Comment by SeaRise** [comment]
/hold
merge after code freeze

**Pull request merged**
